include/deal.II/base/mpi.h

```diff
@@ -1481,21 +1481,28 @@ namespace Utilities
     {
 #  ifndef DEAL_II_WITH_MPI
       (void)comm;
-      Assert(objects_to_send.size() == 0,
+      Assert(objects_to_send.size() < 2,
              ExcMessage("Cannot send to more than one processor."));
       Assert(objects_to_send.find(0) != objects_to_send.end() ||
                objects_to_send.size() == 0,
              ExcMessage("Can only send to myself or to nobody."));
       return objects_to_send;
 #  else
+      const auto my_proc = this_mpi_process(comm);
 
-      std::vector<unsigned int> send_to(objects_to_send.size());
-      {
-        unsigned int i = 0;
-        for (const auto &m : objects_to_send)
-          send_to[i++] = m.first;
-      }
-      AssertDimension(send_to.size(), objects_to_send.size());
+      std::map<unsigned int, T> received_objects;
+
+      std::vector<unsigned int> send_to;
+      send_to.reserve(objects_to_send.size());
+      for (const auto &m : objects_to_send)
+        if (m.first == my_proc)
+          received_objects[my_proc] = m.second;
+        else
+          send_to.emplace_back(m.first);
+
+      // Shortcut, when running in serial
+      if (send_to.size() == 0)
+        return received_objects;
 
       const auto receive_from =
         Utilities::MPI::compute_point_to_point_communication_pattern(comm,
```

Comment:
Only the number of messages to receive is required from this function later on (l.1544). Hence, `Utilities::MPI::compute_n_point_to_point_communications` should suffice here.